cbmc/proofs/proof_guide.md

```diff
@@ -0,0 +1,399 @@
+```
+
+but how can j be both larger-than-or-equal-to 0 AND less-than-or-equal-to -1 at the same time? Answer: it can't! So.. the left hand side of the quantified predicate is False, so it reduces to:
+
+```
+FORALL { int j; false ==> dst[j] == 0 }
+```
+
+The `==>` is a logical implication, and we know that `False ==> ANYTHING` is True, so all is well.
+
+This comes up a lot. You often end up reasoning about such "slices" of arrays, where one or more of the slices has "null range" at either the loop entry, or at the loop exit point, and therefore that particular quantifier "disappears". Several examples of this trick can be found on the MLKEM codebase.
+
+This also explains why we prefer to use _signed_ integers for loop counters and quantified variables - to allow "0 - 1" to evaluate to -1.  If you use an unsigned type, then "0 - 1" evaluated to something like UINT_MAX, and all hell breaks loose.
+
+## Invariant && loop-exit ==> post-condition
+
+Another important sanity check. If there are no statements following the loop body, then the loop invariant AND the loop exit condition should imply the post-condition. For the example above, we need to prove:
+
+```
+// Loop invariant
+(i >= 0 &&
+ i <= len &&
+ (FORALL { int j; (0 <= j && j <= i - 1) ==> dst[j] == 0 } )
+&&
+// Loop exit condition must be TRUE, so
+i == len)
+
+  ==>
+
+// Post-condition
+FORALL { int k; (0 <= k && k < len) ==> dst[k] == 0 }
+```
+
+The loop exit condition means that we can just replace `i` with `len` in the hypotheses, to get:
+
+```
+len >= 0 &&
+len <= len &&
+(FORALL { int j; (0 <= j && j <= len - 1) ==> dst[j] == 0 } )
+  ==>
+FORALL { int k; (0 <= k && k < len) ==> dst[k] == 0 }
+```
+
+`j <= len - 1` can be simplified to `j < len` so that simplifies to:
+
+```
+FORALL { int j; (0 <= j && j < len) ==> dst[j] == 0 }
+  ==>
+FORALL { int k; (0 <= k && k < len) ==> dst[k] == 0 }
+```
+
+which is True.
+
+## Recipe to prove a new function
+
+Let's say we want to develop a proof of a function. Here are the basic steps.
+1. Populate a proof directory
+2. Update Makefile
+3. Update harness function
+4. Supply top-level contracts for the function
+5. Supply loop-invariants (if requied) and other interior contracts
+6. Prove it!
+
+These steps are expanded on in the following sub-sections
+
+### Populate a proof directory
+
+For MLKEM-C-AArch64, proof directories lie below `cbmc/proofs`
+
+Create a new sub-directory in there, where the name of the directory is the name of the function _without_ the `$(KYBER_NAMESPACE)` prefix.
+
+That directory needs to contain 3 files.
+
+* cbmc-proof.txt
+* Makefile
+* XXX_harness.c
+
+where "XXX" is the name of the function being proved - same as the directory name.
+
+I suggest that you copy these files from an existing proof directory and modify the latter two. The `cbmc-proof.txt` file is just a marker that says "this directory contains a CBMC proof" to the tools, so no modification is required.
```

Comment:
Consider adding a template directory?

Reply:
If we add a template directory under "proofs", then the run-cbmc-proofs.py script will spot it and try to run the Makefile in there... this might cause a spurious failure.